make/deploy.mk

```
@@ -0,0 +1,110 @@
+# On arm64, you can force the use of amd64 images:
+#
+#   make deploy CRI_ARCH=amd64
+#   make kind-load CRI_ARCH=amd64
+#
+CRI_ARCH ?= $(HOST_ARCH)
+
+# The commas in FEATURE_GATES have been kept for backwards compatibility with
+# our previous Bazel/scripts.
+FEATURE_GATES ?= AdditionalCertificateOutputFormats=true,ExperimentalCertificateSigningRequestControllers=true,ExperimentalGatewayAPISupport=true,ServerSideApply=true
+
+# In make, there is no way to escape commas or spaces. So we use the variables
+# $(space) and $(comma) instead.
+null  =
+space = $(null) #
+comma = ,
+
+FEATURE_GATES_CONTROLLER = $(subst $(space),$(comma),$(filter AllAlpha=% AllBeta=% AdditionalCertificateOutputFormats=% ValidateCAA=% ExperimentalCertificateSigningRequestControllers=% ExperimentalGatewayAPISupport=% ServerSideApply=%, $(subst $(comma),$(space),$(FEATURE_GATES))))
+FEATURE_GATES_WEBHOOK = $(subst $(space),$(comma),$(filter AllAlpha=% AllBeta=% AdditionalCertificateOutputFormats=% , $(subst $(comma),$(space),$(FEATURE_GATES))))
+FEATURE_GATES_CAINJECTOR = $(subst $(space),$(comma),$(filter AllAlpha=% AllBeta=%, $(subst $(comma),$(space),$(FEATURE_GATES))))
+
+# generated by ./hack/latest-kind-images.sh
+KIND_IMAGE_SHA_K8S_1.18=sha256:f4bcc97a0ad6e7abaf3f643d890add7efe6ee4ab90baeb374b4f41a4c95567eb
+KIND_IMAGE_SHA_K8S_1.19=sha256:a70639454e97a4b733f9d9b67e12c01f6b0297449d5b9cbbef87473458e26dca
+KIND_IMAGE_SHA_K8S_1.20=sha256:cbeaf907fc78ac97ce7b625e4bf0de16e3ea725daf6b04f930bd14c67c671ff9
+KIND_IMAGE_SHA_K8S_1.21=sha256:0fda882e43d425622f045b492f8bd83c2e0b4984fc03e2e05ec101ca1a685fb7
+KIND_IMAGE_SHA_K8S_1.22=sha256:f240c00ffb1d82a2a2225ca0f5c85d1c45aa2b97921327cb3f6da4eee7eae5c3
+KIND_IMAGE_SHA_K8S_1.23=sha256:49824ab1727c04e56a21a5d8372a402fcd32ea51ac96a2706a12af38934f81ac
+
+K8S_VERSION = 1.23
+
+# Creates a Kind cluster and wait for all nodes to be ready.
+KIND_CLUSTER_NAME ?= kind
+.PHONY: kind-cluster
+kind-cluster: devel/cluster/config/v1beta2.yaml | ${KIND} ${KUBECTL}
+	@if [ -z "${KIND_IMAGE_SHA_K8S_$(K8S_VERSION)}" ]; then echo "this K8S_VERSION is not supported"; exit 1; fi
+	@${KIND} get clusters | grep -q ${KIND_CLUSTER_NAME} || \
+		${KIND} create cluster --config $< --name ${KIND_CLUSTER_NAME} --wait 5m --image docker.io/kindest/node@${KIND_IMAGE_SHA_K8S_$(K8S_VERSION)}
+	@if test "$$(kubectl config current-context 2>/dev/null)" != kind-${KIND_CLUSTER_NAME}; then \
+		printf "$(Y)$(WARN)Warning$(E): your current kubectl context isn't equal to kind-${KIND_CLUSTER_NAME}. Run the following command:\n" >&2; \
+		printf "    $(C)kubectl config use-context kind-${KIND_CLUSTER_NAME}$(E)\n" >&2; \
+		exit 1; \
+	fi
```

suggestion (non-blocking): I think the logic here is complicated enough that I'd like to see it in its own script file, so that the target becomes a call to a bash script. I don't think it's a blocker at all though, as long as the logic works the same way in the makefile as it did in the old scripts.

The same would apply to other PHONY targets in this file; I think they cross the threshold of being complicated enough for their own scripts. But it's not a big deal!

```suggestion
.PHONY: kind-cluster
kind-cluster: devel/cluster/config/v1beta2.yaml | ${KIND} ${KUBECTL}
	./devel/cluster/create-kind.sh ${KIND} ${KUBECTL} # the name of the script doesn't really matter
```